[12.0] Add graph of dependencies between jobs

guewen opened the pull request:
Issue described in #129.

# Introduction

This PR adds a way to express dependencies between jobs, so called jobs graphs.

The delaying API has been rewritten in a way that allows to express these dependencies. Under the hood, an acyclic direct graph is built and stored in the job records. Jobs which depend on another or several other jobs are in a state `wait_dependencies`, when a job is done, it finds the dependent jobs, which is set to `pending` if all the other dependent jobs are done as well.

# API changes

## Single jobs

The new API is entirely backward compatible. Delaying a single job is done as before with:

```python
records.with_delay(**job_options).my_method(*args, **kwargs)
```

A new concept of delayable is introduced to compose graph, the above is equivalent to:

```python
delayable = records.delayable(**job_options)
delayable.my_method(*args, **kwargs)
delayable.delay()
```

## Chains

A very simple graph, execute a job when the first is done, can be written as:

```python
initial_delayable = records.delayable(**job_options)
initial_delayable.my_method(*args, **kwargs)
initial_delayable.on_done(records.delayable().my_other_method())
initial_delayable.delay()
```

`.delay()` is called only on the root of the graph.

If several jobs are chained, it can be easier to build a chain which are executed in sequence:

```python
from odoo.addons.queue_job.delay import chain

chain(
    records.delayable().my_method(),
    records.delayable().my_second_method(),
   records.delayable().my_third_method(),
).delay()
```

## Groups

Dependencies permit depending on a group of jobs, the third method is executed when the two jobs of the group are done:

```python
from odoo.addons.queue_job.delay import group

group(
    records.delayable().my_method(),
   records.delayable().my_second_method(),
).on_done(records.delayable().my_third_method()).delay()
```

# UI for graphs

Changes for the graphs and some various tweaks:

* The jobs of a graph all share a "Graph UUID". We can search and group jobs per Graph UUID.
* From the form view of a job, we can quickly open all the other jobs of the same graph
* A new widget shows a visualization of the graph in the form view
* The UUID, graph UUID, enqueued time and func_string are now shown only in debug mode

# Improved testing

## Unit tests

A new testing facility is added, which helps a lot to run assertions on enqueued jobs. It is strongly encouraged to write unit tests using this assert. It works on any jobs, this is not specially related to graphs.

Inside the `mock_jobs` context manager, jobs are never stored in database. However, they are kept in memory in the context manager instance, which allow running checks on them.

```python
    # tests
    from odoo.addons.queue_job.tests.common import mock_jobs


    def test_method_to_test(self):
        with mock_jobs() as jobs_tester:
            result = self.env["model"].method_to_test()
            expected_count = 12

            jobs_tester.assert_jobs_count(1, only=self.env["model"].my_job_method)
            jobs_tester.assert_enqueued_job(
                self.env["model"].my_job_method,
                args=("Hi!",),
                kwargs=dict(count=expected_count),
                properties=dict(priority=15)
            )
            self.assertEqual(result, expected_count)
```

![Output of failed assertion](https://user-images.githubusercontent.com/417223/126867483-ab3172b1-77f8-430e-ba55-21ee2376f578.png)

Additionally, as the jobs are stored in the context manager instance, we can even choose to execute them in the test or not, depending on what we are testing.

```python
             jobs_tester.perform_enqueued_jobs()
```

## Test jobs

The controller `/queue_job/create_test_job` can now generate a random graph of jobs, using the parameters:

* priority: priority of the created jobs
* max_retries: max_retries of the created jobs
* channel: channel of the created jobs
* description: you guess what
* size: number of jobs delayed, if 1, it will not be a graph
* failure_rate: between 0 and 1, if > 0 jobs will randomly fail according to this rate

# Warnings

As of today, there is no correct handling of identity keys for graphs. We still need to figure out the most correct way to handle an identity key for a graph. However, the identity key still work the same way for single jobs.

Closes #129

Commit:
Store dependencies

Commit:
Add wait dependencies state

Commit:
Enqueue waiting jobs when parent jobs are done

Commit:
Optimize and make enqueue of waiting jobs more robust

Commit:
Adapt views for state wait_dependencies

Commit:
Add API for Delayables

Commit:
Fix tests failing when test_queue_job is installed

test_queue_job will create channels root.sub and root.sub.sub, so the
tests doing the same thing will fail with a unique constraint error.

Commit:
Add widget to show job dependencies on UI

Commit:
Show the dependency widget in a tab

When displayed in a tab, the widget show the nodes offset, the d3
network is probably confused by the size or visibility of its canvas.
Install a listener on tabs to trigger a fit() on the network.

Commit:
Add documentation on 'base' model public methods

Commit:
Improve loading of dependencies using batch read

Commit:
Use Delayable in DelayableRecordset

guewen commented on queue_job/delay.py:
```diff
@@ -130,9 +131,31 @@ def delay(self):
+        # If all the jobs of the graph have another job with the same identity,
+        # we do not create them. Maybe we should check that the found jobs are
+        # part of the same graph, but not sure it's really required...
+        # Also, maybe we want to check only the root jobs.
```
idea: add an uuid field for the whole graph (improve search as well)

guewen commented:
I pushed new commits. I still have lots of docs to write and not everything is ready yet, but I would welcome feedback by people interested in the feature :) You can look at the doc of the 'base' model methods and at the tests to have an idea of the dependency API.

guewen commented on queue_job/job.py:
```diff
@@ -244,13 +196,25 @@ class Job(object):
+    def load_many(cls, env, job_uuids):
+        """Read jobs in batch from the Database
+
+        Jobs not found are ignored.
```
todo: log a warning

guewen commented on queue_job/models/queue_job.py:
```diff
@@ -103,6 +107,53 @@ def init(self):
+                reverse_depends_on = dependencies.get(
+                    'reverse_depends_on', []
+                )
+                parents = self.search([
```
Using a single uuid for the whole graph should improve this a lot.

guewen commented:
Something I'm really not sure about is how to deal with identity keys. Currently, I'm checking that if a graph contains some jobs with an identity key, if all the jobs in the graph with an identity key have another job in the db with the same key, then the whole graph is discarded. It doesn't seem very intuitive. What would make most sense to me is to have a single identity key for the whole graph. I'll think about how we could implement this.

guewen commented:
Remaining

* use a single uuid for identifying a graph
* handling of identity keys (not really defined yet)
* handle deprecation of a few things
* implement graph traversal to execute job synchronously when TEST_QUEUE_JOB_NO_DELAY is used
* take care of setting the proper state on requeued children jobs when their parents are not done
* docs
* extend test coverage (I don't remember what is/isn't tested)
* migration to 13.0
* wish: performance analysis, if time allows...
* no "Result" will be handled yet, it can be added later

guewen committed:
Add documentation

guewen committed:
Add a graph UUID

A graph of jobs always share the same graph_uuid, which can be
used to group jobs, but is also a faster way to find all the jobs
of the graph. Then we can use the dependencies field to build the
whole graph from the pre-selection of jobs.

guewen committed:
Hide some technical fields

guewen committed:
Ignore requeues on dependency jobs waiting on parent jobs

Jobs waiting that their dependencies are executed cannot be requeued.
They have to keep waiting their turn.

guewen committed:
Fix lint

guewen committed:
Fix warnings in tests

As jobs are created in the setup, but not always delayed,
the __del__ warning that the job was not delayed is triggered.
Create only the jobs necessary for each test.

guewen committed:
Update vis-network js

guewen committed:
Improve display of jobs graph widget

guewen committed:
Add powerful context manager for running tests on jobs

* mock_jobs context manager with assert methods to check enqueued jobs
* jobs are not stored in database, yet they can be inspected and performed
  during the test
* the context manager works with graphs enqueued by delayable and with with_delay()
* direct execution of jobs works with graphs by executing them synchronously
  following their topological sort

guewen committed:
Set graph_uuid only once in DelayableGraph

The previous implementation was incomplete, as it was setting
the graph_uuid when `Job.add_depends()` was called. It had the
advantage of being correct when (for e.g. a test) manually created
a Job and called `add_depends()`, but it could not work when in
some situations involving a group where one of the job has no job
depending on it.

Also, remove ``Job.add_reverse_depends()`` which is not used.

guewen committed:
Add docstrings on the new delayable classes

guewen committed:
Add option to generate a graph in create_test_job controller

guewen committed:
Improve graph widget

* Fix a js bug when there is no graph.
* Use a single read group for multiple jobs

guewen committed:
Add a smart button to open all the jobs of a graph

guewen committed:
Fix duplicate label

guewen committed:
Fix graph widget now showing title as HTML

From https://visjs.github.io/vis-network/examples/network/other/html-in-titles.html

guewen committed:
Improve graph widget performance

When showing large graphs, like 300 jobs (not really a real life case maybe),
the graph takes a dozen of seconds to render. Disabling the stabilization
helps a lot.

etobella commented:
@guewen Do you need some help with this?

guewen committed:
Escape strings passed to the graph js widget

guewen commented:
@etobella The only thing yet to do I think is about handling the identity key for a whole graph. Well, there is an implementation now which is that if all jobs have an identity key and all these identity keys already exist, we skip the graph. I don't think this is how it should be done, rather have a single identity key for the whole graph.

Anyway, I think this could be handled later, the "how" could become more clear at usage.

This PR is in a safe enough state to be tested if you'd like :) I should open my PR soon, I'd like to improve the description first 😆

guewen committed:
Fix lint

simahawk commented:
Beautiful work, as usual :)
Question: any chance to add this feature to a separated module?
It's quite big as change and it will be easier to maintain (and to maintain the core) if it stays separated.

guewen commented:
> Question: any chance to add this feature to a separated module?

This is not something "on top" of the existing API, but an improvement of the current API to support graphs. Enqueuing a single job is now (aside a few details) enqueueing a graph of 1 job, the mechanism is generic for supporting both (`with_delay` goes through the graph machinery - syntactic sugar when you want only 1 job -, but the performance is the same as today when you have 1 job).  In other words: this is the core now ;)

guewen commented on queue_job/views/queue_job_views.xml:
```diff
@@ -26,16 +26,27 @@
+                            name="open_graph_jobs"
+                            type="object"
+                            class="oe_stat_button"
+                            icon="fa-pencil-square-o"
```
todo: find a better icon

etobella replied:
On Fontawesome 5 I would propose https://fontawesome.com/v5.15/icons/project-diagram?style=solid

With Fontawesome 4.7 it is hard to find something, as there is not many options :disappointed:

gregory-moka replied:
Maybe : https://fontawesome.com/v4.7/icon/sitemap

guewen replied:
The sitemap is great, thanks!

guewen commented:
Anyone interested in this?
The PR is ready ;)

sbidoul reviewed (approved):
Fantastic work @guewen ! This is going to be extremely useful.

Had you considered 'then' instead of 'done' ?

A word in the docs about what happens when a job of graph fails might be helpful.

guewen commented:
Thanks @sbidoul! 
It has been a work on the long run as I did the first commit on the 2nd of July 2019 and the last commits on the 2nd of November 2021 :)

> Had you considered 'then' instead of 'done' ?

I remember I did! But I don't remember why I preferred `done`. But if I think about it again, I would say that in the jquery promises (where I probably took the inspiration?), `then` will be called for both success and failures, and `done` only for success, and I wanted to emphasize on this. Nowadays, jquery is less and less used and I don't think promises have a `done` function.

Do you think it would be better with `then`? It would be an easy change that I'll be happy to do if it makes the API better.

> A word in the docs about what happens when a job of graph fails might be helpful.

Yep, I'll do that.

BTW, I just pushed the 14.0 version here https://github.com/OCA/queue/pull/403

And while I'm here, @sbidoul, outstanding work on runboat :green_heart:

guewen committed:
Add documentation about handling of failures in a graph

sbidoul commented:
> Do you think it would be better with then? It would be an easy change that I'll be happy to do if it makes the API better.

My first impression was that `done()` could create some confusion with `set_done()`.
And Indeed js promises have `then()`, `catch()` and `finally()`. Although js `then` accepts a second argument for then rejected case. Not sure... up to you :)

simahawk commented:
Could be done more verbose like `on_done` and `on_fail`.
Regarding using `then` I'm not sure it will add clarity :)
But yeah... not a game changer in this wonderful work :)

@guewen do you prefer to merge 12.0 first?

guewen committed:
Rename mock_jobs() to trap_jobs()

The fact that mocks are used is an implementation detail, and is actually
hidden to the end user of the test API.

guewen commented:
`on_done()` is maybe a better choice, it is less ambiguous than `then()` (clearer on the fact it happens on `done()`) and `done()` (less confusion with `set_done()`). I'll do the renaming soon, unless you have other suggestions.

I was thinking about the naming of `mock_jobs()` too, and renamed it to `trap_jobs()`, because I think the latter describes better what *it does*, vs *how it does it* for the former (which is a hidden technical detail).

guewen commented:
> @guewen do you prefer to merge 12.0 first?

I suppose I'll do the last renaming changes here, then cherry-pick them on 14.0, so yes it'd make sense.

guewen committed:
Rename dependency method done() to on_done()

on_done() is maybe a better choice, it is less ambiguous than then() (clearer
on the fact it happens on done()) and done() (less confusion with set_done()).
I'll do the renaming soon, unless you have other suggestions.

Ref: https://github.com/OCA/queue/pull/154#issuecomment-1025127060

guewen commented:
/ocabot merge major

simahawk commented:
seems it failed... 
BTW are you planning to fwd port?

/ocabot merge major

sbidoul commented:
Apparently travis did not report back ? I take the opportunity to test the freshly merged OCA repo tempalte for 12.0 in #426, and move to GitHub actions.

sbidoul commented:
Ok, done. @guewen can you rebase ?

guewen commented:
/ocabot merge major

guewen commented:
@simahawk version for 14.0 is there: https://github.com/OCA/queue/pull/403
I didn't work on any other version yet.
